apm-agent-core/src/test/java/co/elastic/apm/agent/impl/transaction/TraceContextTest.java

```diff
@@ -473,11 +481,109 @@ void testRandomValue() {
     void testSetSampled() {
         final TraceContext traceContext = TraceContext.with64BitId(tracer);
         traceContext.asRootSpan(ConstantSampler.of(false));
+
+        // not sampled means zero sample rate
         assertThat(traceContext.isSampled()).isFalse();
+        assertThat(traceContext.getSampleRate()).isEqualTo(0d);
+
+        // sampled without sample rate
         traceContext.setRecorded(true);
+
         assertThat(traceContext.isSampled()).isTrue();
+        assertThat(traceContext.getSampleRate()).isNull();
+
+        // sampled with sample rate
+        traceContext.getTraceState().setSampleRate(0.5d);
+
+        assertThat(traceContext.isSampled()).isTrue();
+        assertThat(traceContext.getSampleRate()).isEqualTo(0.5d);
+
+        // not sampled, sample rate should be unset
         traceContext.setRecorded(false);
         assertThat(traceContext.isSampled()).isFalse();
+        assertThat(traceContext.getSampleRate()).isEqualTo(0.0d);
+    }
+
+    @Test
+    void testRootSpanShouldAddsSampleRateToTraceState() {
+        final TraceContext traceContext = createRootSpan(0.42d);
+        String traceState = traceContext.getTraceState().toTextHeader(Integer.MAX_VALUE);
+        assertThat(traceState).isEqualTo("es=s:0.42");
+    }
+
+    private TraceContext createRootSpan(double sampleRate){
+        final TraceContext traceContext = TraceContext.with64BitId(tracer);
+
+        Sampler sampler = mock(Sampler.class);
+        when(sampler.isSampled(any(Id.class))).thenReturn(true);
+        when(sampler.getSampleRate()).thenReturn(sampleRate);
+
+        traceContext.asRootSpan(sampler);
+        return traceContext;
+    }
+
+    @Test
+    void testTracedChildSpanWithoutTraceState() {
+        Map<String, String> headers = Map.of(
+            TraceContext.W3C_TRACE_PARENT_TEXTUAL_HEADER_NAME, "00-0af7651916cd43dd8448eb211c80319c-b9c7c989f97918e1-01"
+        );
+        TraceContext child = createChildSpanFromHeaders(headers);
+
+        assertThat(child.isSampled()).isTrue();
+        assertThat(child.getSampleRate()).isNull();
+    }
+
+    @Test
+    void testNonTracedChildSpanWithoutTraceState() {
+        Map<String, String> headers = Map.of(
+            TraceContext.W3C_TRACE_PARENT_TEXTUAL_HEADER_NAME, "00-0af7651916cd43dd8448eb211c80319c-b9c7c989f97918e1-00"
+        );
+        TraceContext child = createChildSpanFromHeaders(headers);
+
+        assertThat(child.isSampled()).isFalse();
+        assertThat(child.getSampleRate()).isEqualTo(0.0d);
+    }
+
+    @ParameterizedTest
+    @CsvSource(delimiter = '|', value = {
+        // invalid tracestate values: just assume no sample rate is provided
+        "|null",
+        "es=|null",
+        "es=s|null",
+        "es=s:|null",
+        "es=s:a|null",
+        // valid tracestate values with sample rate
+        "es=s:1|1",
+        "es=s:0.42|0.42",
```

Comment:
```suggestion
        "es=s:0.42|0.42",
        "es=d:0.42|(maybe 1.0?)",
```

Reply:
I looked quickly at ruby (ref impl) and go agents, and both consider invalid values implies same behavior as if they were simply absent.

Reply:
OK, I guess other tests cover that